crates/egui/src/atomics/atomic_layout.rs

```diff
@@ -0,0 +1,392 @@
+                    .find(|a| matches!(a.kind, AtomicKind::Text(..)));
+                if let Some(atomic) = first_text {
+                    atomic.shrink = true;
+                }
+            }
+        }
+
+        let id = id.unwrap_or_else(|| ui.next_auto_id());
+
+        let fallback_text_color =
+            fallback_text_color.unwrap_or_else(|| ui.style().visuals.text_color());
+        let gap = gap.unwrap_or(ui.spacing().icon_spacing);
+
+        // The size available for the content
+        let available_inner_size = ui.available_size() - frame.total_margin().sum();
+
+        let mut desired_width = 0.0;
+        let mut preferred_width = 0.0;
+        let mut preferred_height = 0.0;
+
+        let mut height: f32 = 0.0;
+
+        let mut sized_items = Vec::new();
+
+        let mut grow_count = 0;
+
+        let mut shrink_item = None;
+
+        let align2 = align2.unwrap_or_else(|| {
+            Align2([ui.layout().horizontal_align(), ui.layout().vertical_align()])
+        });
+
+        if atomics.len() > 1 {
+            let gap_space = gap * (atomics.len() as f32 - 1.0);
+            desired_width += gap_space;
+            preferred_width += gap_space;
+        }
+
+        let default_font_height = || {
+            let font_selection = FontSelection::default();
+            let font_id = font_selection.resolve(ui.style());
+            ui.fonts(|f| f.row_height(&font_id))
+        };
+
+        let max_font_size = ui
+            .fonts(|fonts| {
+                atomics
+                    .iter()
+                    .filter_map(|a| a.get_min_height_for_image(fonts, ui.style()))
+                    .max_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal))
+            })
+            .unwrap_or_else(default_font_height);
+
+        for (idx, item) in atomics.into_iter().enumerate() {
+            if item.shrink {
+                debug_assert!(
+                    shrink_item.is_none(),
+                    "Only one atomic may be marked as shrink"
+                );
+                if shrink_item.is_none() {
+                    shrink_item = Some((idx, item));
+                    continue;
+                }
+            }
+            if item.grow {
+                grow_count += 1;
+            }
+            let sized = item.into_sized(ui, available_inner_size, max_font_size, Some(wrap_mode));
+            let size = sized.size;
+
+            desired_width += size.x;
+            preferred_width += sized.preferred_size.x;
+
+            height = height.at_least(size.y);
+            preferred_height = preferred_height.at_least(sized.preferred_size.y);
+
+            sized_items.push(sized);
+        }
+
+        if let Some((index, item)) = shrink_item {
+            // The `shrink` item gets the remaining space
+            let shrunk_size = Vec2::new(
+                available_inner_size.x - desired_width,
+                available_inner_size.y,
+            );
```

Comment:
This is confusing. It looks like `desired_width` does not include the size of the `shrink` item.

What you are calling `shrunk_size` seems to be how much _extra_ space we have for X (which should be distributed to the _grow_ item, not the _shrink_ item)